File: swift-algorithm/swift-algorithm/LeetCode/LeetCode/1_two_sum.swift
//
//  main.swift
//  swift-algorithm
//
//  Created by 전판근 on 2020/12/28.
//

import Foundation

//MARK:- 시간 복잡도 O(n^2)
class Solution {
    func twoSum(_ nums: [Int], _ target: Int) -> [Int] {
        for i in 0..<nums.count {
            for j in i+1..<nums.count {
                if (nums[i] + nums[j] == target) {
                    return [i, j]
                }
            }
        }
        return []
    }
}


//MARK:- 시간 복잡도 O(n) HashTable 사용
class Solution {
    func twoSum(_ nums: [Int], _ target: Int) -> [Int] {
        var nums_map: Dictionary = [Int:Int]()
        
        for i in 0..<nums.count {
            let ans1 = nums[i]
            let ans2 = target - ans1
            
            if let j = nums_map[ans2] { return [j, i] }
            nums_map[ans1] = i
        }
        
        return []
    }
}

File: python/LeetCode/1_two_sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_maps = {}

        # 키와 값을 바꿔서 딕셔너리로 저장
        for i, num in enumerate(nums):
            nums_maps[num] = i

        for i, num in enumerate(nums):
            if target - num in nums_maps and i != nums_maps[target - num]:
                return nums.index(num), nums_maps[target - num]
